// @tencent/polaris/dist/plugins/naming/polaris-server/connector/embedded.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.envAwaredDiscover = exports.kDiscoverHosts = exports.DiscoverBackend = exports.normalizeEmbeddedHosts = void 0;
const normalizeEmbeddedHosts = (address, port) => Object.freeze(address.map(host => `${host}:${port}`));
exports.normalizeEmbeddedHosts = normalizeEmbeddedHosts;
var DiscoverBackend;
(function (DiscoverBackend) {
    /** 正式环境 */
    DiscoverBackend[DiscoverBackend["Formal"] = 0] = "Formal";
    /** 测试环境 */
    DiscoverBackend[DiscoverBackend["Debug"] = 1] = "Debug";
})(DiscoverBackend = exports.DiscoverBackend || (exports.DiscoverBackend = {}));
exports.kDiscoverHosts = Object.freeze({
    [DiscoverBackend.Debug]: [
        "9.205.2.29",
        "9.205.2.123"
    ],
    [DiscoverBackend.Formal]: [
        "9.141.66.244",
        "9.146.202.27",
        "9.146.200.81",
        "9.146.202.35",
        "9.146.205.191",
        "9.146.200.61",
        "9.141.66.219",
        "9.141.65.110",
        "9.141.65.29",
        "9.141.121.7"
    ]
});
const envAwaredDiscover = () => {
    if (process.env.POLARIS_NAMING_DISCOVER_DEBUG) {
        return DiscoverBackend.Debug;
    }
    return DiscoverBackend.Formal;
};
exports.envAwaredDiscover = envAwaredDiscover;
