// @tencent/taf-stream/dev/test/map/longTestProxy_Tag@readwirte.js
// **********************************************************************
// This file was generated by a TAF parser!
// TAF version 3.2.2.11 by WSRD Tencent.
// Generated from `/data/NodejsServer/NodejsDeployServer/jce/uploads/1506320805573/in/longTest.jce'
// **********************************************************************
/*global assert*/
"use strict";

var TafStream = require("../../../");
//var TafError  = require("@tencent/taf-rpc").error;

var LONGTEST = LONGTEST || {};
module.exports.LONGTEST = LONGTEST;


LONGTEST.MapLong = function() {
    this.m1 = new TafStream.Map(TafStream.String, TafStream.Int64);
    this.m2 = new TafStream.Map(TafStream.Int64, TafStream.String);
    this.m3 = new TafStream.Map(TafStream.String, TafStream.Map(TafStream.String, TafStream.Int64));
    this.m4 = new TafStream.Map(TafStream.String, TafStream.Map(TafStream.Int64, TafStream.String));
    this.m5 = new TafStream.List(TafStream.Map(TafStream.String, TafStream.Int64));
    this._classname = "LONGTEST.MapLong";
};
LONGTEST.MapLong._classname = "LONGTEST.MapLong";
LONGTEST.MapLong._write = function (os, tag, value) { os.writeStruct(tag, value); }
LONGTEST.MapLong._read  = function (is, tag, def) { return is.readStruct(tag, true, def); }
LONGTEST.MapLong._readFrom = function (is) {
    var tmp = new LONGTEST.MapLong();
    tmp.m1 = is.readMap(0, false, TafStream.Map(TafStream.String, TafStream.Int64), false, true);
    tmp.m2 = is.readMap(1, false, TafStream.Map(TafStream.Int64, TafStream.String), false, true);
    tmp.m3 = is.readMap(2, false, TafStream.Map(TafStream.String, TafStream.Map(TafStream.String, TafStream.Int64)), false, true);
    tmp.m4 = is.readMap(3, false, TafStream.Map(TafStream.String, TafStream.Map(TafStream.Int64, TafStream.String)), false, true);
    tmp.m5 = is.readList(4, false, TafStream.List(TafStream.Map(TafStream.String, TafStream.Int64)), false, true);
    return tmp;
};
LONGTEST.MapLong.prototype._writeTo = function (os) {
    os.writeMap(0, this.m1, false, true);
    os.writeMap(1, this.m2, false, true);
    os.writeMap(2, this.m3, false, true);
    os.writeMap(3, this.m4, false, true);
    os.writeList(4, this.m5, false, true);
};
LONGTEST.MapLong.prototype._equal = function () {
    assert.fail("this structure not define key operation");
}
LONGTEST.MapLong.prototype._genKey = function () {
    if (!this._proto_struct_name_) {
        this._proto_struct_name_ = "STRUCT" + Math.random();
    }
    return this._proto_struct_name_;
}
LONGTEST.MapLong.prototype.toObject = function() { 
    return {
        "m1" : this.m1.toObject(),
        "m2" : this.m2.toObject(),
        "m3" : this.m3.toObject(),
        "m4" : this.m4.toObject(),
        "m5" : this.m5.toObject()
    };
}
LONGTEST.MapLong.prototype.readFromObject = function(json) { 
    json.hasOwnProperty("m1") && (this.m1.readFromObject(json.m1));
    json.hasOwnProperty("m2") && (this.m2.readFromObject(json.m2));
    json.hasOwnProperty("m3") && (this.m3.readFromObject(json.m3));
    json.hasOwnProperty("m4") && (this.m4.readFromObject(json.m4));
    json.hasOwnProperty("m5") && (this.m5.readFromObject(json.m5));
}
LONGTEST.MapLong.prototype.toBinBuffer = function () {
    var os = new TafStream.JceOutputStream();
    this._writeTo(os);
    return os.getBinBuffer();
}
LONGTEST.MapLong.new = function () {
    return new LONGTEST.MapLong();
}
LONGTEST.MapLong.create = function (is) {
    return LONGTEST.MapLong._readFrom(is);
}
