// @tencent/polaris/dist/plugins/logging/console.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConsoleTraceLogging = void 0;
const debug_1 = require("debug");
const logging_1 = require("../../logging");
const plugins_1 = require("../../plugins");
class ConsoleTraceLogging {
    constructor() {
        this.type = plugins_1.PluginType.TraceLogging;
        this.name = "ConsoleTraceLogging";
        this.tracers = Object.create(null); // 使用 `protected` 便于继承时重写
    }
    log(severity, prefix, message, ...optionalParams) {
        const logger = console[logging_1.LogVerbosity[severity].toLowerCase()];
        if (prefix !== undefined) {
            logger(prefix, message, ...optionalParams);
        }
        else {
            logger(message, ...optionalParams);
        }
    }
    trace(tracer, transaction, message, ...optionalParams) {
        let trace = this.tracers[tracer];
        if (!trace) {
            trace = (0, debug_1.default)(tracer);
            this.tracers[tracer] = trace;
        }
        if (transaction !== undefined) {
            trace(`<${transaction}>`, message, ...optionalParams);
        }
        else {
            trace(message, ...optionalParams);
        }
    }
}
exports.ConsoleTraceLogging = ConsoleTraceLogging;
