// @tencent/taf-stream/dev/flatTest/NewFlatDecodeTestProxy.js
// **********************************************************************
// This file was generated by a TAF parser!
// TAF version 3.2.2.11 by WSRD Tencent.
// Generated from `/data/NodejsServer/NodejsDeployServer/jce/uploads/1508723390063/in/DecodeTest.jce'
// **********************************************************************

"use strict";

var assert    = require("assert");
var TafStream = require("../../libjce");

var DecodeTest = DecodeTest || {};
module.exports.DecodeTest = DecodeTest;


DecodeTest.TestStruct = function(inputStream) {
    this._inputStream = inputStream;
    this._originPosition = 0;
    this._int1 = undefined;
    this._int2 = undefined;
    this._str1 = undefined;
    this._str2 = undefined;
    this._classname = "DecodeTest.TestStruct";
};
DecodeTest.TestStruct.prototype.__defineGetter__("int1", function () {
    if(this._int1 == undefined){
        if(this._inputStream == undefined){
            return 0;
        }
        var currentPosition = this._inputStream._binBuffer._position;
        this._inputStream._binBuffer._position = this._originPosition;
        this._int1 = this._inputStream.readInt32(0, false, 0);
        this._inputStream._binBuffer._position = currentPosition;
    }
    return this._int1;
});
DecodeTest.TestStruct.prototype.__defineSetter__("int1", function (value) {
    this._int1 = value;
});

DecodeTest.TestStruct.prototype.__defineGetter__("int2", function () {
    if(this._int2 == undefined){
        if(this._inputStream == undefined){
            return 0;
        }
        var currentPosition = this._inputStream._binBuffer._position;
        this._inputStream._binBuffer._position = this._originPosition;
        this._int2 = this._inputStream.readInt32(1, false, 0);
        this._inputStream._binBuffer._position = currentPosition;
    }
    return this._int2;
});
DecodeTest.TestStruct.prototype.__defineSetter__("int2", function (value) {
    this._int2 = value;
});

DecodeTest.TestStruct.prototype.__defineGetter__("str1", function () {
    if(this._str1 == undefined){
        if(this._inputStream == undefined){
            return "";
        }
        var currentPosition = this._inputStream._binBuffer._position;
        this._inputStream._binBuffer._position = this._originPosition;
        this._str1 = this._inputStream.readString(2, false, "");
        this._inputStream._binBuffer._position = currentPosition;
    }
    return this._str1;
});
DecodeTest.TestStruct.prototype.__defineSetter__("str1", function (value) {
    this._str1 = value;
});

DecodeTest.TestStruct.prototype.__defineGetter__("str2", function () {
    if(this._str2 == undefined){
        if(this._inputStream == undefined){
            return "";
        }
        var currentPosition = this._inputStream._binBuffer._position;
        this._inputStream._binBuffer._position = this._originPosition;
        this._str2 = this._inputStream.readString(3, false, "");
        this._inputStream._binBuffer._position = currentPosition;
    }
    return this._str2;
});
DecodeTest.TestStruct.prototype.__defineSetter__("str2", function (value) {
    this._str2 = value;
});


DecodeTest.TestStruct._classname = "DecodeTest.TestStruct";
DecodeTest.TestStruct._write = function (os, tag, value) { os.writeStruct(tag, value); }
DecodeTest.TestStruct._read  = function (is, tag, def) { return is.readStruct(tag, true, def); }
DecodeTest.TestStruct._readFrom = function (is) {
    var tmp = new DecodeTest.TestStruct(is);
    return tmp;
};
DecodeTest.TestStruct.prototype._writeTo = function (os) {
    os.writeInt32(0, this.int1);
    os.writeInt32(1, this.int2);
    os.writeString(2, this.str1);
    os.writeString(3, this.str2);
};
DecodeTest.TestStruct.prototype._equal = function () {
    assert.fail("this structure not define key operation");
}
DecodeTest.TestStruct.prototype._genKey = function () {
    if (!this._proto_struct_name_) {
        this._proto_struct_name_ = "STRUCT" + Math.random();
    }
    return this._proto_struct_name_;
}
DecodeTest.TestStruct.prototype.toObject = function() { 
    return {
        "int1" : this.int1,
        "int2" : this.int2,
        "str1" : this.str1,
        "str2" : this.str2
    };
}
DecodeTest.TestStruct.prototype.readFromObject = function(json) { 
    json.hasOwnProperty("int1") && (this.int1 = json.int1);
    json.hasOwnProperty("int2") && (this.int2 = json.int2);
    json.hasOwnProperty("str1") && (this.str1 = json.str1);
    json.hasOwnProperty("str2") && (this.str2 = json.str2);
}
DecodeTest.TestStruct.prototype.toBinBuffer = function () {
    var os = new TafStream.JceOutputStream();
    this._writeTo(os);
    return os.getBinBuffer();
}
DecodeTest.TestStruct.new = function () {
    return new DecodeTest.TestStruct();
}
DecodeTest.TestStruct.create = function (is) {
    return DecodeTest.TestStruct._readFrom(is);
}
