// @tencent/toggle-server-node/lib/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataTypeJSON = exports.DataTypeList = exports.DataTypeFloat = exports.DataTypeInt = exports.DataTypeBool = exports.DataTypeString = exports.DataTypeMap = exports.WithDataset = exports.WithoutDataset = exports.TabReportEventCode = exports.TabExperimentCode = exports.TabPercentage = exports.TabToggleOn = exports.TabToggleOff = exports.TabExperimentCodeField = exports.TabToggleNameField = exports.TabToggleValField = exports.AttaToken = exports.AttaId = exports.FALSE_STR = exports.TRUE_STR = exports.MATCH_TYPE_REG = exports.MATCH_TYPE_EQUAL = exports.BLACK_LIST = exports.WHITE_LIST = exports.FEATURE_TIME_LIMIT_TYPE_FIXED = exports.FEATURE_TIME_LIMIT_TYPE_DAILY = exports.FEATURE_TIME_LIMIT_TYPE_NONE = exports.STG_RELATION_OR = exports.STG_RELATION_AND = exports.FEATURE_VALUE_STR_FALSE = exports.FEATURE_VALUE_STR_TRUE = exports.SET_MIGRATION_CODE = exports.HTTP_SERVER_ERROR_CODE = exports.HTTP_NOT_FOUND_CODE = exports.HTTP_UNAUTHORIZED_CODE = exports.HTTP_OK_CODE = exports.MAX_HTTP_RETRY = void 0;
exports.MAX_HTTP_RETRY = 3;
exports.HTTP_OK_CODE = 200;
exports.HTTP_UNAUTHORIZED_CODE = 401;
exports.HTTP_NOT_FOUND_CODE = 404;
exports.HTTP_SERVER_ERROR_CODE = 500;
exports.SET_MIGRATION_CODE = 100302;
exports.FEATURE_VALUE_STR_TRUE = 'true';
exports.FEATURE_VALUE_STR_FALSE = 'false';
exports.STG_RELATION_AND = 2;
exports.STG_RELATION_OR = 1;
exports.FEATURE_TIME_LIMIT_TYPE_NONE = 1;
exports.FEATURE_TIME_LIMIT_TYPE_DAILY = 2;
exports.FEATURE_TIME_LIMIT_TYPE_FIXED = 3;
exports.WHITE_LIST = 'whitelist';
exports.BLACK_LIST = 'blacklist';
exports.MATCH_TYPE_EQUAL = 'equal';
exports.MATCH_TYPE_REG = 'regexp';
exports.TRUE_STR = 'true';
exports.FALSE_STR = 'false';
exports.AttaId = '03d00037385';
exports.AttaToken = '3868159371';
exports.TabToggleValField = '_feature_flag_status';
exports.TabToggleNameField = '_feature_flag_key';
exports.TabExperimentCodeField = 'experimentCode';
exports.TabToggleOff = '0';
exports.TabToggleOn = '1';
exports.TabPercentage = 'percentage';
exports.TabExperimentCode = 'experimentCode';
exports.TabReportEventCode = 'ft_experiment';
exports.WithoutDataset = 0;
exports.WithDataset = 1;
exports.DataTypeMap = 'MAP';
exports.DataTypeString = 'STRING';
exports.DataTypeBool = 'BOOL';
exports.DataTypeInt = 'INT';
exports.DataTypeFloat = 'FLOAT';
exports.DataTypeList = 'LIST';
exports.DataTypeJSON = 'JSON';
