// @tencent/toggle-server-node/lib/dtoJce.js
"use strict";
// **********************************************************************
// Parsed By TafParser(3.2.2.20), Generated By jce2node(20200804)
// TafParser Maintained By <TAF> and jce2node Maintained By <superzheng>
// Generated from "dto.jce" by Structure Mode
// **********************************************************************
Object.defineProperty(exports, "__esModule", { value: true });
exports.dto = void 0;
/* tslint:disable */
/* eslint-disable */
/// <reference types="node" />
var assert = require("assert");
var TafStream = require("@tencent/taf-stream");
var _hasOwnProperty = Object.prototype.hasOwnProperty;
var dto;
(function (dto) {
    dto.jceProtocolVersion = "1.6.0";
    var ClientContext = /** @class */ (function () {
        function ClientContext() {
            this.userId = "";
            this.sessionId = "";
            this.remoteAddress = "";
            this.deviceId = "";
            this.hostname = "";
            this.properties = new TafStream.Map(TafStream.String, TafStream.String);
            this.qimei = "";
            this._proto_struct_name_ = "";
            this._classname = "dto.ClientContext";
        }
        ClientContext._write = function (os, tag, val) { os.writeStruct(tag, val); };
        ClientContext._read = function (is, tag, def) { return is.readStruct(tag, true, def); };
        ClientContext._readFrom = function (is) {
            var tmp = new ClientContext;
            tmp.userId = is.readString(0, false, "");
            tmp.sessionId = is.readString(1, false, "");
            tmp.remoteAddress = is.readString(2, false, "");
            tmp.deviceId = is.readString(3, false, "");
            tmp.hostname = is.readString(4, false, "");
            tmp.properties = is.readMap(5, false, TafStream.Map(TafStream.String, TafStream.String));
            tmp.qimei = is.readString(6, false, "");
            return tmp;
        };
        ClientContext.prototype._writeTo = function (os) {
            os.writeString(0, this.userId);
            os.writeString(1, this.sessionId);
            os.writeString(2, this.remoteAddress);
            os.writeString(3, this.deviceId);
            os.writeString(4, this.hostname);
            os.writeMap(5, this.properties);
            os.writeString(6, this.qimei);
        };
        ClientContext.prototype._equal = function () {
            assert.fail("this structure not define key operation");
        };
        ClientContext.prototype._genKey = function () {
            if (!this._proto_struct_name_) {
                this._proto_struct_name_ = "STRUCT" + Math.random();
            }
            return this._proto_struct_name_;
        };
        ClientContext.prototype.toObject = function () {
            return {
                userId: this.userId,
                sessionId: this.sessionId,
                remoteAddress: this.remoteAddress,
                deviceId: this.deviceId,
                hostname: this.hostname,
                properties: this.properties.toObject(),
                qimei: this.qimei
            };
        };
        ClientContext.prototype.readFromObject = function (json) {
            _hasOwnProperty.call(json, "userId") && (this.userId = json.userId);
            _hasOwnProperty.call(json, "sessionId") && (this.sessionId = json.sessionId);
            _hasOwnProperty.call(json, "remoteAddress") && (this.remoteAddress = json.remoteAddress);
            _hasOwnProperty.call(json, "deviceId") && (this.deviceId = json.deviceId);
            _hasOwnProperty.call(json, "hostname") && (this.hostname = json.hostname);
            _hasOwnProperty.call(json, "properties") && (this.properties.readFromObject(json.properties));
            _hasOwnProperty.call(json, "qimei") && (this.qimei = json.qimei);
            return this;
        };
        ClientContext.prototype.toBinBuffer = function () {
            var os = new TafStream.JceOutputStream();
            this._writeTo(os);
            return os.getBinBuffer();
        };
        ClientContext.new = function () {
            return new ClientContext();
        };
        ClientContext.create = function (is) {
            return dto.ClientContext._readFrom(is);
        };
        ClientContext._classname = "dto.ClientContext";
        return ClientContext;
    }());
    dto.ClientContext = ClientContext;
    var ExtendInfo = /** @class */ (function () {
        function ExtendInfo() {
            this.bundleId = "";
            this.qua = "";
            this.channel = "";
            this.os = "";
            this.osVersion = "";
            this.sdkVersion = "";
            this.properties = new TafStream.Map(TafStream.String, TafStream.String);
            this._proto_struct_name_ = "";
            this._classname = "dto.ExtendInfo";
        }
        ExtendInfo._write = function (os, tag, val) { os.writeStruct(tag, val); };
        ExtendInfo._read = function (is, tag, def) { return is.readStruct(tag, true, def); };
        ExtendInfo._readFrom = function (is) {
            var tmp = new ExtendInfo;
            tmp.bundleId = is.readString(0, false, "");
            tmp.qua = is.readString(1, false, "");
            tmp.channel = is.readString(2, false, "");
            tmp.os = is.readString(3, false, "");
            tmp.osVersion = is.readString(4, false, "");
            tmp.sdkVersion = is.readString(5, false, "");
            tmp.properties = is.readMap(6, false, TafStream.Map(TafStream.String, TafStream.String));
            return tmp;
        };
        ExtendInfo.prototype._writeTo = function (os) {
            os.writeString(0, this.bundleId);
            os.writeString(1, this.qua);
            os.writeString(2, this.channel);
            os.writeString(3, this.os);
            os.writeString(4, this.osVersion);
            os.writeString(5, this.sdkVersion);
            os.writeMap(6, this.properties);
        };
        ExtendInfo.prototype._equal = function () {
            assert.fail("this structure not define key operation");
        };
        ExtendInfo.prototype._genKey = function () {
            if (!this._proto_struct_name_) {
                this._proto_struct_name_ = "STRUCT" + Math.random();
            }
            return this._proto_struct_name_;
        };
        ExtendInfo.prototype.toObject = function () {
            return {
                bundleId: this.bundleId,
                qua: this.qua,
                channel: this.channel,
                os: this.os,
                osVersion: this.osVersion,
                sdkVersion: this.sdkVersion,
                properties: this.properties.toObject()
            };
        };
        ExtendInfo.prototype.readFromObject = function (json) {
            _hasOwnProperty.call(json, "bundleId") && (this.bundleId = json.bundleId);
            _hasOwnProperty.call(json, "qua") && (this.qua = json.qua);
            _hasOwnProperty.call(json, "channel") && (this.channel = json.channel);
            _hasOwnProperty.call(json, "os") && (this.os = json.os);
            _hasOwnProperty.call(json, "osVersion") && (this.osVersion = json.osVersion);
            _hasOwnProperty.call(json, "sdkVersion") && (this.sdkVersion = json.sdkVersion);
            _hasOwnProperty.call(json, "properties") && (this.properties.readFromObject(json.properties));
            return this;
        };
        ExtendInfo.prototype.toBinBuffer = function () {
            var os = new TafStream.JceOutputStream();
            this._writeTo(os);
            return os.getBinBuffer();
        };
        ExtendInfo.new = function () {
            return new ExtendInfo();
        };
        ExtendInfo.create = function (is) {
            return dto.ExtendInfo._readFrom(is);
        };
        ExtendInfo._classname = "dto.ExtendInfo";
        return ExtendInfo;
    }());
    dto.ExtendInfo = ExtendInfo;
    var QueryFeatureReq = /** @class */ (function () {
        function QueryFeatureReq() {
            this.productId = "";
            this.moduleId = "";
            this.moduleVersion = "";
            this.clientCtx = new dto.ClientContext;
            this.extendInfo = new dto.ExtendInfo;
            this.timestamp = 0;
            this.localIdList = new TafStream.List(TafStream.Int32);
            this._proto_struct_name_ = "";
            this._classname = "dto.QueryFeatureReq";
        }
        QueryFeatureReq._write = function (os, tag, val) { os.writeStruct(tag, val); };
        QueryFeatureReq._read = function (is, tag, def) { return is.readStruct(tag, true, def); };
        QueryFeatureReq._readFrom = function (is) {
            var tmp = new QueryFeatureReq;
            tmp.productId = is.readString(0, true, "");
            tmp.moduleId = is.readString(1, true, "");
            tmp.moduleVersion = is.readString(2, true, "");
            tmp.clientCtx = is.readStruct(3, false, dto.ClientContext);
            tmp.extendInfo = is.readStruct(4, false, dto.ExtendInfo);
            tmp.timestamp = is.readInt64(5, false, 0);
            tmp.localIdList = is.readList(6, false, TafStream.List(TafStream.Int32));
            return tmp;
        };
        QueryFeatureReq.prototype._writeTo = function (os) {
            os.writeString(0, this.productId);
            os.writeString(1, this.moduleId);
            os.writeString(2, this.moduleVersion);
            os.writeStruct(3, this.clientCtx);
            os.writeStruct(4, this.extendInfo);
            os.writeInt64(5, this.timestamp);
            os.writeList(6, this.localIdList);
        };
        QueryFeatureReq.prototype._equal = function () {
            assert.fail("this structure not define key operation");
        };
        QueryFeatureReq.prototype._genKey = function () {
            if (!this._proto_struct_name_) {
                this._proto_struct_name_ = "STRUCT" + Math.random();
            }
            return this._proto_struct_name_;
        };
        QueryFeatureReq.prototype.toObject = function () {
            return {
                productId: this.productId,
                moduleId: this.moduleId,
                moduleVersion: this.moduleVersion,
                clientCtx: this.clientCtx.toObject(),
                extendInfo: this.extendInfo.toObject(),
                timestamp: this.timestamp,
                localIdList: this.localIdList.toObject()
            };
        };
        QueryFeatureReq.prototype.readFromObject = function (json) {
            _hasOwnProperty.call(json, "productId") && (this.productId = json.productId);
            _hasOwnProperty.call(json, "moduleId") && (this.moduleId = json.moduleId);
            _hasOwnProperty.call(json, "moduleVersion") && (this.moduleVersion = json.moduleVersion);
            _hasOwnProperty.call(json, "clientCtx") && (this.clientCtx.readFromObject(json.clientCtx));
            _hasOwnProperty.call(json, "extendInfo") && (this.extendInfo.readFromObject(json.extendInfo));
            _hasOwnProperty.call(json, "timestamp") && (this.timestamp = json.timestamp);
            _hasOwnProperty.call(json, "localIdList") && (this.localIdList.readFromObject(json.localIdList));
            return this;
        };
        QueryFeatureReq.prototype.toBinBuffer = function () {
            var os = new TafStream.JceOutputStream();
            this._writeTo(os);
            return os.getBinBuffer();
        };
        QueryFeatureReq.new = function () {
            return new QueryFeatureReq();
        };
        QueryFeatureReq.create = function (is) {
            return dto.QueryFeatureReq._readFrom(is);
        };
        QueryFeatureReq._classname = "dto.QueryFeatureReq";
        return QueryFeatureReq;
    }());
    dto.QueryFeatureReq = QueryFeatureReq;
    var TimeLimit = /** @class */ (function () {
        function TimeLimit() {
            this.startTime = "";
            this.endTime = "";
            this._proto_struct_name_ = "";
            this._classname = "dto.TimeLimit";
        }
        TimeLimit._write = function (os, tag, val) { os.writeStruct(tag, val); };
        TimeLimit._read = function (is, tag, def) { return is.readStruct(tag, true, def); };
        TimeLimit._readFrom = function (is) {
            var tmp = new TimeLimit;
            tmp.startTime = is.readString(0, false, "");
            tmp.endTime = is.readString(1, false, "");
            return tmp;
        };
        TimeLimit.prototype._writeTo = function (os) {
            os.writeString(0, this.startTime);
            os.writeString(1, this.endTime);
        };
        TimeLimit.prototype._equal = function () {
            assert.fail("this structure not define key operation");
        };
        TimeLimit.prototype._genKey = function () {
            if (!this._proto_struct_name_) {
                this._proto_struct_name_ = "STRUCT" + Math.random();
            }
            return this._proto_struct_name_;
        };
        TimeLimit.prototype.toObject = function () {
            return {
                startTime: this.startTime,
                endTime: this.endTime
            };
        };
        TimeLimit.prototype.readFromObject = function (json) {
            _hasOwnProperty.call(json, "startTime") && (this.startTime = json.startTime);
            _hasOwnProperty.call(json, "endTime") && (this.endTime = json.endTime);
            return this;
        };
        TimeLimit.prototype.toBinBuffer = function () {
            var os = new TafStream.JceOutputStream();
            this._writeTo(os);
            return os.getBinBuffer();
        };
        TimeLimit.new = function () {
            return new TimeLimit();
        };
        TimeLimit.create = function (is) {
            return dto.TimeLimit._readFrom(is);
        };
        TimeLimit._classname = "dto.TimeLimit";
        return TimeLimit;
    }());
    dto.TimeLimit = TimeLimit;
    var DatasetEntity = /** @class */ (function () {
        function DatasetEntity() {
            this.datasetId = 0;
            this.versionId = 0;
            this.data = new TafStream.Map(TafStream.String, TafStream.String);
            this.weight = 0;
            this.type = "";
            this.value = "";
            this._proto_struct_name_ = "";
            this._classname = "dto.DatasetEntity";
        }
        DatasetEntity._write = function (os, tag, val) { os.writeStruct(tag, val); };
        DatasetEntity._read = function (is, tag, def) { return is.readStruct(tag, true, def); };
        DatasetEntity._readFrom = function (is) {
            var tmp = new DatasetEntity;
            tmp.datasetId = is.readInt32(0, false, 0);
            tmp.versionId = is.readInt32(1, false, 0);
            tmp.data = is.readMap(2, false, TafStream.Map(TafStream.String, TafStream.String));
            tmp.weight = is.readInt32(3, false, 0);
            tmp.type = is.readString(4, false, "");
            tmp.value = is.readString(5, false, "");
            return tmp;
        };
        DatasetEntity.prototype._writeTo = function (os) {
            os.writeInt32(0, this.datasetId);
            os.writeInt32(1, this.versionId);
            os.writeMap(2, this.data);
            os.writeInt32(3, this.weight);
            os.writeString(4, this.type);
            os.writeString(5, this.value);
        };
        DatasetEntity.prototype._equal = function () {
            assert.fail("this structure not define key operation");
        };
        DatasetEntity.prototype._genKey = function () {
            if (!this._proto_struct_name_) {
                this._proto_struct_name_ = "STRUCT" + Math.random();
            }
            return this._proto_struct_name_;
        };
        DatasetEntity.prototype.toObject = function () {
            return {
                datasetId: this.datasetId,
                versionId: this.versionId,
                data: this.data.toObject(),
                weight: this.weight,
                type: this.type,
                value: this.value
            };
        };
        DatasetEntity.prototype.readFromObject = function (json) {
            _hasOwnProperty.call(json, "datasetId") && (this.datasetId = json.datasetId);
            _hasOwnProperty.call(json, "versionId") && (this.versionId = json.versionId);
            _hasOwnProperty.call(json, "data") && (this.data.readFromObject(json.data));
            _hasOwnProperty.call(json, "weight") && (this.weight = json.weight);
            _hasOwnProperty.call(json, "type") && (this.type = json.type);
            _hasOwnProperty.call(json, "value") && (this.value = json.value);
            return this;
        };
        DatasetEntity.prototype.toBinBuffer = function () {
            var os = new TafStream.JceOutputStream();
            this._writeTo(os);
            return os.getBinBuffer();
        };
        DatasetEntity.new = function () {
            return new DatasetEntity();
        };
        DatasetEntity.create = function (is) {
            return dto.DatasetEntity._readFrom(is);
        };
        DatasetEntity._classname = "dto.DatasetEntity";
        return DatasetEntity;
    }());
    dto.DatasetEntity = DatasetEntity;
    var FeatureResult = /** @class */ (function () {
        function FeatureResult() {
            this.id = 0;
            this.name = "";
            this.result = "";
            this.extendField = new TafStream.Map(TafStream.String, TafStream.String);
            this.isAbtFirst = true;
            this.timeLimitType = 1;
            this.timeLimits = new TafStream.List(dto.TimeLimit);
            this.refreshType = 1;
            this.dataset = new dto.DatasetEntity;
            this._proto_struct_name_ = "";
            this._classname = "dto.FeatureResult";
        }
        FeatureResult._write = function (os, tag, val) { os.writeStruct(tag, val); };
        FeatureResult._read = function (is, tag, def) { return is.readStruct(tag, true, def); };
        FeatureResult._readFrom = function (is) {
            var tmp = new FeatureResult;
            tmp.id = is.readInt32(0, false, 0);
            tmp.name = is.readString(1, false, "");
            tmp.result = is.readString(2, false, "");
            tmp.extendField = is.readMap(3, false, TafStream.Map(TafStream.String, TafStream.String));
            tmp.isAbtFirst = is.readBoolean(4, false, true);
            tmp.timeLimitType = is.readInt32(5, false, 1);
            tmp.timeLimits = is.readList(6, false, TafStream.List(dto.TimeLimit));
            tmp.refreshType = is.readInt32(7, false, 1);
            tmp.dataset = is.readStruct(8, false, dto.DatasetEntity);
            return tmp;
        };
        FeatureResult.prototype._writeTo = function (os) {
            os.writeInt32(0, this.id);
            os.writeString(1, this.name);
            os.writeString(2, this.result);
            os.writeMap(3, this.extendField);
            os.writeBoolean(4, this.isAbtFirst);
            os.writeInt32(5, this.timeLimitType);
            os.writeList(6, this.timeLimits);
            os.writeInt32(7, this.refreshType);
            os.writeStruct(8, this.dataset);
        };
        FeatureResult.prototype._equal = function () {
            assert.fail("this structure not define key operation");
        };
        FeatureResult.prototype._genKey = function () {
            if (!this._proto_struct_name_) {
                this._proto_struct_name_ = "STRUCT" + Math.random();
            }
            return this._proto_struct_name_;
        };
        FeatureResult.prototype.toObject = function () {
            return {
                id: this.id,
                name: this.name,
                result: this.result,
                extendField: this.extendField.toObject(),
                isAbtFirst: this.isAbtFirst,
                timeLimitType: this.timeLimitType,
                timeLimits: this.timeLimits.toObject(),
                refreshType: this.refreshType,
                dataset: this.dataset.toObject()
            };
        };
        FeatureResult.prototype.readFromObject = function (json) {
            _hasOwnProperty.call(json, "id") && (this.id = json.id);
            _hasOwnProperty.call(json, "name") && (this.name = json.name);
            _hasOwnProperty.call(json, "result") && (this.result = json.result);
            _hasOwnProperty.call(json, "extendField") && (this.extendField.readFromObject(json.extendField));
            _hasOwnProperty.call(json, "isAbtFirst") && (this.isAbtFirst = json.isAbtFirst);
            _hasOwnProperty.call(json, "timeLimitType") && (this.timeLimitType = json.timeLimitType);
            _hasOwnProperty.call(json, "timeLimits") && (this.timeLimits.readFromObject(json.timeLimits));
            _hasOwnProperty.call(json, "refreshType") && (this.refreshType = json.refreshType);
            _hasOwnProperty.call(json, "dataset") && (this.dataset.readFromObject(json.dataset));
            return this;
        };
        FeatureResult.prototype.toBinBuffer = function () {
            var os = new TafStream.JceOutputStream();
            this._writeTo(os);
            return os.getBinBuffer();
        };
        FeatureResult.new = function () {
            return new FeatureResult();
        };
        FeatureResult.create = function (is) {
            return dto.FeatureResult._readFrom(is);
        };
        FeatureResult._classname = "dto.FeatureResult";
        return FeatureResult;
    }());
    dto.FeatureResult = FeatureResult;
    var QueryFeatureResp = /** @class */ (function () {
        function QueryFeatureResp() {
            this.code = 0;
            this.msg = "";
            this.nextTime = 0;
            this.setName = "";
            this.data = new TafStream.List(dto.FeatureResult);
            this.timestamp = 0;
            this.deletedFeatures = new TafStream.List(TafStream.Int32);
            this.increment = true;
            this.serverTime = 0;
            this._proto_struct_name_ = "";
            this._classname = "dto.QueryFeatureResp";
        }
        QueryFeatureResp._write = function (os, tag, val) { os.writeStruct(tag, val); };
        QueryFeatureResp._read = function (is, tag, def) { return is.readStruct(tag, true, def); };
        QueryFeatureResp._readFrom = function (is) {
            var tmp = new QueryFeatureResp;
            tmp.code = is.readInt32(0, true, 0);
            tmp.msg = is.readString(1, true, "");
            tmp.nextTime = is.readInt32(2, false, 0);
            tmp.setName = is.readString(3, false, "");
            tmp.data = is.readList(4, false, TafStream.List(dto.FeatureResult));
            tmp.timestamp = is.readInt64(5, false, 0);
            tmp.deletedFeatures = is.readList(6, false, TafStream.List(TafStream.Int32));
            tmp.increment = is.readBoolean(7, false, true);
            tmp.serverTime = is.readInt64(8, false, 0);
            return tmp;
        };
        QueryFeatureResp.prototype._writeTo = function (os) {
            os.writeInt32(0, this.code);
            os.writeString(1, this.msg);
            os.writeInt32(2, this.nextTime);
            os.writeString(3, this.setName);
            os.writeList(4, this.data);
            os.writeInt64(5, this.timestamp);
            os.writeList(6, this.deletedFeatures);
            os.writeBoolean(7, this.increment);
            os.writeInt64(8, this.serverTime);
        };
        QueryFeatureResp.prototype._equal = function () {
            assert.fail("this structure not define key operation");
        };
        QueryFeatureResp.prototype._genKey = function () {
            if (!this._proto_struct_name_) {
                this._proto_struct_name_ = "STRUCT" + Math.random();
            }
            return this._proto_struct_name_;
        };
        QueryFeatureResp.prototype.toObject = function () {
            return {
                code: this.code,
                msg: this.msg,
                nextTime: this.nextTime,
                setName: this.setName,
                data: this.data.toObject(),
                timestamp: this.timestamp,
                deletedFeatures: this.deletedFeatures.toObject(),
                increment: this.increment,
                serverTime: this.serverTime
            };
        };
        QueryFeatureResp.prototype.readFromObject = function (json) {
            _hasOwnProperty.call(json, "code") && (this.code = json.code);
            _hasOwnProperty.call(json, "msg") && (this.msg = json.msg);
            _hasOwnProperty.call(json, "nextTime") && (this.nextTime = json.nextTime);
            _hasOwnProperty.call(json, "setName") && (this.setName = json.setName);
            _hasOwnProperty.call(json, "data") && (this.data.readFromObject(json.data));
            _hasOwnProperty.call(json, "timestamp") && (this.timestamp = json.timestamp);
            _hasOwnProperty.call(json, "deletedFeatures") && (this.deletedFeatures.readFromObject(json.deletedFeatures));
            _hasOwnProperty.call(json, "increment") && (this.increment = json.increment);
            _hasOwnProperty.call(json, "serverTime") && (this.serverTime = json.serverTime);
            return this;
        };
        QueryFeatureResp.prototype.toBinBuffer = function () {
            var os = new TafStream.JceOutputStream();
            this._writeTo(os);
            return os.getBinBuffer();
        };
        QueryFeatureResp.new = function () {
            return new QueryFeatureResp();
        };
        QueryFeatureResp.create = function (is) {
            return dto.QueryFeatureResp._readFrom(is);
        };
        QueryFeatureResp._classname = "dto.QueryFeatureResp";
        return QueryFeatureResp;
    }());
    dto.QueryFeatureResp = QueryFeatureResp;
    var Strategy = /** @class */ (function () {
        function Strategy() {
            this.name = "";
            this.parameters = new TafStream.Map(TafStream.String, TafStream.String);
            this._proto_struct_name_ = "";
            this._classname = "dto.Strategy";
        }
        Strategy._write = function (os, tag, val) { os.writeStruct(tag, val); };
        Strategy._read = function (is, tag, def) { return is.readStruct(tag, true, def); };
        Strategy._readFrom = function (is) {
            var tmp = new Strategy;
            tmp.name = is.readString(0, false, "");
            tmp.parameters = is.readMap(1, false, TafStream.Map(TafStream.String, TafStream.String));
            return tmp;
        };
        Strategy.prototype._writeTo = function (os) {
            os.writeString(0, this.name);
            os.writeMap(1, this.parameters);
        };
        Strategy.prototype._equal = function () {
            assert.fail("this structure not define key operation");
        };
        Strategy.prototype._genKey = function () {
            if (!this._proto_struct_name_) {
                this._proto_struct_name_ = "STRUCT" + Math.random();
            }
            return this._proto_struct_name_;
        };
        Strategy.prototype.toObject = function () {
            return {
                name: this.name,
                parameters: this.parameters.toObject()
            };
        };
        Strategy.prototype.readFromObject = function (json) {
            _hasOwnProperty.call(json, "name") && (this.name = json.name);
            _hasOwnProperty.call(json, "parameters") && (this.parameters.readFromObject(json.parameters));
            return this;
        };
        Strategy.prototype.toBinBuffer = function () {
            var os = new TafStream.JceOutputStream();
            this._writeTo(os);
            return os.getBinBuffer();
        };
        Strategy.new = function () {
            return new Strategy();
        };
        Strategy.create = function (is) {
            return dto.Strategy._readFrom(is);
        };
        Strategy._classname = "dto.Strategy";
        return Strategy;
    }());
    dto.Strategy = Strategy;
    var Prerequisite = /** @class */ (function () {
        function Prerequisite() {
            this.versionId = 0;
            this.statusValue = "";
            this._proto_struct_name_ = "";
            this._classname = "dto.Prerequisite";
        }
        Prerequisite._write = function (os, tag, val) { os.writeStruct(tag, val); };
        Prerequisite._read = function (is, tag, def) { return is.readStruct(tag, true, def); };
        Prerequisite._readFrom = function (is) {
            var tmp = new Prerequisite;
            tmp.versionId = is.readInt32(0, false, 0);
            tmp.statusValue = is.readString(1, false, "");
            return tmp;
        };
        Prerequisite.prototype._writeTo = function (os) {
            os.writeInt32(0, this.versionId);
            os.writeString(1, this.statusValue);
        };
        Prerequisite.prototype._equal = function () {
            assert.fail("this structure not define key operation");
        };
        Prerequisite.prototype._genKey = function () {
            if (!this._proto_struct_name_) {
                this._proto_struct_name_ = "STRUCT" + Math.random();
            }
            return this._proto_struct_name_;
        };
        Prerequisite.prototype.toObject = function () {
            return {
                versionId: this.versionId,
                statusValue: this.statusValue
            };
        };
        Prerequisite.prototype.readFromObject = function (json) {
            _hasOwnProperty.call(json, "versionId") && (this.versionId = json.versionId);
            _hasOwnProperty.call(json, "statusValue") && (this.statusValue = json.statusValue);
            return this;
        };
        Prerequisite.prototype.toBinBuffer = function () {
            var os = new TafStream.JceOutputStream();
            this._writeTo(os);
            return os.getBinBuffer();
        };
        Prerequisite.new = function () {
            return new Prerequisite();
        };
        Prerequisite.create = function (is) {
            return dto.Prerequisite._readFrom(is);
        };
        Prerequisite._classname = "dto.Prerequisite";
        return Prerequisite;
    }());
    dto.Prerequisite = Prerequisite;
    var DatasetAssign = /** @class */ (function () {
        function DatasetAssign() {
            this.toggleValue = true;
            this.datasetType = 0;
            this.dataset = new dto.DatasetEntity;
            this.rolloutBase = "";
            this.rolloutWeights = new TafStream.List(dto.DatasetEntity);
            this._proto_struct_name_ = "";
            this._classname = "dto.DatasetAssign";
        }
        DatasetAssign._write = function (os, tag, val) { os.writeStruct(tag, val); };
        DatasetAssign._read = function (is, tag, def) { return is.readStruct(tag, true, def); };
        DatasetAssign._readFrom = function (is) {
            var tmp = new DatasetAssign;
            tmp.toggleValue = is.readBoolean(0, false, true);
            tmp.datasetType = is.readInt32(1, false, 0);
            tmp.dataset = is.readStruct(2, false, dto.DatasetEntity);
            tmp.rolloutBase = is.readString(3, false, "");
            tmp.rolloutWeights = is.readList(4, false, TafStream.List(dto.DatasetEntity));
            return tmp;
        };
        DatasetAssign.prototype._writeTo = function (os) {
            os.writeBoolean(0, this.toggleValue);
            os.writeInt32(1, this.datasetType);
            os.writeStruct(2, this.dataset);
            os.writeString(3, this.rolloutBase);
            os.writeList(4, this.rolloutWeights);
        };
        DatasetAssign.prototype._equal = function () {
            assert.fail("this structure not define key operation");
        };
        DatasetAssign.prototype._genKey = function () {
            if (!this._proto_struct_name_) {
                this._proto_struct_name_ = "STRUCT" + Math.random();
            }
            return this._proto_struct_name_;
        };
        DatasetAssign.prototype.toObject = function () {
            return {
                toggleValue: this.toggleValue,
                datasetType: this.datasetType,
                dataset: this.dataset.toObject(),
                rolloutBase: this.rolloutBase,
                rolloutWeights: this.rolloutWeights.toObject()
            };
        };
        DatasetAssign.prototype.readFromObject = function (json) {
            _hasOwnProperty.call(json, "toggleValue") && (this.toggleValue = json.toggleValue);
            _hasOwnProperty.call(json, "datasetType") && (this.datasetType = json.datasetType);
            _hasOwnProperty.call(json, "dataset") && (this.dataset.readFromObject(json.dataset));
            _hasOwnProperty.call(json, "rolloutBase") && (this.rolloutBase = json.rolloutBase);
            _hasOwnProperty.call(json, "rolloutWeights") && (this.rolloutWeights.readFromObject(json.rolloutWeights));
            return this;
        };
        DatasetAssign.prototype.toBinBuffer = function () {
            var os = new TafStream.JceOutputStream();
            this._writeTo(os);
            return os.getBinBuffer();
        };
        DatasetAssign.new = function () {
            return new DatasetAssign();
        };
        DatasetAssign.create = function (is) {
            return dto.DatasetAssign._readFrom(is);
        };
        DatasetAssign._classname = "dto.DatasetAssign";
        return DatasetAssign;
    }());
    dto.DatasetAssign = DatasetAssign;
    var StrategyGroup = /** @class */ (function () {
        function StrategyGroup() {
            this.id = 0;
            this.stgRelation = 1;
            this.strategies = new TafStream.List(dto.Strategy);
            this.datasetEx = new dto.DatasetAssign;
            this._proto_struct_name_ = "";
            this._classname = "dto.StrategyGroup";
        }
        StrategyGroup._write = function (os, tag, val) { os.writeStruct(tag, val); };
        StrategyGroup._read = function (is, tag, def) { return is.readStruct(tag, true, def); };
        StrategyGroup._readFrom = function (is) {
            var tmp = new StrategyGroup;
            tmp.id = is.readInt32(0, false, 0);
            tmp.stgRelation = is.readInt32(1, false, 1);
            tmp.strategies = is.readList(2, false, TafStream.List(dto.Strategy));
            tmp.datasetEx = is.readStruct(3, false, dto.DatasetAssign);
            return tmp;
        };
        StrategyGroup.prototype._writeTo = function (os) {
            os.writeInt32(0, this.id);
            os.writeInt32(1, this.stgRelation);
            os.writeList(2, this.strategies);
            os.writeStruct(3, this.datasetEx);
        };
        StrategyGroup.prototype._equal = function () {
            assert.fail("this structure not define key operation");
        };
        StrategyGroup.prototype._genKey = function () {
            if (!this._proto_struct_name_) {
                this._proto_struct_name_ = "STRUCT" + Math.random();
            }
            return this._proto_struct_name_;
        };
        StrategyGroup.prototype.toObject = function () {
            return {
                id: this.id,
                stgRelation: this.stgRelation,
                strategies: this.strategies.toObject(),
                datasetEx: this.datasetEx.toObject()
            };
        };
        StrategyGroup.prototype.readFromObject = function (json) {
            _hasOwnProperty.call(json, "id") && (this.id = json.id);
            _hasOwnProperty.call(json, "stgRelation") && (this.stgRelation = json.stgRelation);
            _hasOwnProperty.call(json, "strategies") && (this.strategies.readFromObject(json.strategies));
            _hasOwnProperty.call(json, "datasetEx") && (this.datasetEx.readFromObject(json.datasetEx));
            return this;
        };
        StrategyGroup.prototype.toBinBuffer = function () {
            var os = new TafStream.JceOutputStream();
            this._writeTo(os);
            return os.getBinBuffer();
        };
        StrategyGroup.new = function () {
            return new StrategyGroup();
        };
        StrategyGroup.create = function (is) {
            return dto.StrategyGroup._readFrom(is);
        };
        StrategyGroup._classname = "dto.StrategyGroup";
        return StrategyGroup;
    }());
    dto.StrategyGroup = StrategyGroup;
    var FlowStatus = /** @class */ (function () {
        function FlowStatus() {
            this.id = 0;
            this.name = "";
            this.ruleType = 0;
            this.strategyGroups = new TafStream.List(dto.StrategyGroup);
            this._proto_struct_name_ = "";
            this._classname = "dto.FlowStatus";
        }
        FlowStatus._write = function (os, tag, val) { os.writeStruct(tag, val); };
        FlowStatus._read = function (is, tag, def) { return is.readStruct(tag, true, def); };
        FlowStatus._readFrom = function (is) {
            var tmp = new FlowStatus;
            tmp.id = is.readInt32(0, false, 0);
            tmp.name = is.readString(1, false, "");
            tmp.ruleType = is.readInt32(2, false, 0);
            tmp.strategyGroups = is.readList(3, false, TafStream.List(dto.StrategyGroup));
            return tmp;
        };
        FlowStatus.prototype._writeTo = function (os) {
            os.writeInt32(0, this.id);
            os.writeString(1, this.name);
            os.writeInt32(2, this.ruleType);
            os.writeList(3, this.strategyGroups);
        };
        FlowStatus.prototype._equal = function () {
            assert.fail("this structure not define key operation");
        };
        FlowStatus.prototype._genKey = function () {
            if (!this._proto_struct_name_) {
                this._proto_struct_name_ = "STRUCT" + Math.random();
            }
            return this._proto_struct_name_;
        };
        FlowStatus.prototype.toObject = function () {
            return {
                id: this.id,
                name: this.name,
                ruleType: this.ruleType,
                strategyGroups: this.strategyGroups.toObject()
            };
        };
        FlowStatus.prototype.readFromObject = function (json) {
            _hasOwnProperty.call(json, "id") && (this.id = json.id);
            _hasOwnProperty.call(json, "name") && (this.name = json.name);
            _hasOwnProperty.call(json, "ruleType") && (this.ruleType = json.ruleType);
            _hasOwnProperty.call(json, "strategyGroups") && (this.strategyGroups.readFromObject(json.strategyGroups));
            return this;
        };
        FlowStatus.prototype.toBinBuffer = function () {
            var os = new TafStream.JceOutputStream();
            this._writeTo(os);
            return os.getBinBuffer();
        };
        FlowStatus.new = function () {
            return new FlowStatus();
        };
        FlowStatus.create = function (is) {
            return dto.FlowStatus._readFrom(is);
        };
        FlowStatus._classname = "dto.FlowStatus";
        return FlowStatus;
    }());
    dto.FlowStatus = FlowStatus;
    var FeatureVersion = /** @class */ (function () {
        function FeatureVersion() {
            this.id = 0;
            this.name = "";
            this.version = "";
            this.value = "";
            this.strategies = new TafStream.List(dto.Strategy);
            this.extendField = new TafStream.Map(TafStream.String, TafStream.String);
            this.isAbtFirst = true;
            this.stgRelation = 1;
            this.timeLimitType = 1;
            this.timeLimits = new TafStream.List(dto.TimeLimit);
            this.refreshType = 1;
            this.prerequisites = new TafStream.List(dto.Prerequisite);
            this.strategyGroups = new TafStream.List(dto.StrategyGroup);
            this.defaultDatasetEx = new dto.DatasetAssign;
            this.offDatasetEx = new dto.DatasetAssign;
            this.flowStatus = new dto.FlowStatus;
            this._proto_struct_name_ = "";
            this._classname = "dto.FeatureVersion";
        }
        FeatureVersion._write = function (os, tag, val) { os.writeStruct(tag, val); };
        FeatureVersion._read = function (is, tag, def) { return is.readStruct(tag, true, def); };
        FeatureVersion._readFrom = function (is) {
            var tmp = new FeatureVersion;
            tmp.id = is.readInt32(0, false, 0);
            tmp.name = is.readString(1, false, "");
            tmp.version = is.readString(2, false, "");
            tmp.value = is.readString(3, false, "");
            tmp.strategies = is.readList(4, false, TafStream.List(dto.Strategy));
            tmp.extendField = is.readMap(5, false, TafStream.Map(TafStream.String, TafStream.String));
            tmp.isAbtFirst = is.readBoolean(6, false, true);
            tmp.stgRelation = is.readInt32(7, false, 1);
            tmp.timeLimitType = is.readInt32(8, false, 1);
            tmp.timeLimits = is.readList(9, false, TafStream.List(dto.TimeLimit));
            tmp.refreshType = is.readInt32(10, false, 1);
            tmp.prerequisites = is.readList(11, false, TafStream.List(dto.Prerequisite));
            tmp.strategyGroups = is.readList(12, false, TafStream.List(dto.StrategyGroup));
            tmp.defaultDatasetEx = is.readStruct(13, false, dto.DatasetAssign);
            tmp.offDatasetEx = is.readStruct(14, false, dto.DatasetAssign);
            tmp.flowStatus = is.readStruct(15, false, dto.FlowStatus);
            return tmp;
        };
        FeatureVersion.prototype._writeTo = function (os) {
            os.writeInt32(0, this.id);
            os.writeString(1, this.name);
            os.writeString(2, this.version);
            os.writeString(3, this.value);
            os.writeList(4, this.strategies);
            os.writeMap(5, this.extendField);
            os.writeBoolean(6, this.isAbtFirst);
            os.writeInt32(7, this.stgRelation);
            os.writeInt32(8, this.timeLimitType);
            os.writeList(9, this.timeLimits);
            os.writeInt32(10, this.refreshType);
            os.writeList(11, this.prerequisites);
            os.writeList(12, this.strategyGroups);
            os.writeStruct(13, this.defaultDatasetEx);
            os.writeStruct(14, this.offDatasetEx);
            os.writeStruct(15, this.flowStatus);
        };
        FeatureVersion.prototype._equal = function () {
            assert.fail("this structure not define key operation");
        };
        FeatureVersion.prototype._genKey = function () {
            if (!this._proto_struct_name_) {
                this._proto_struct_name_ = "STRUCT" + Math.random();
            }
            return this._proto_struct_name_;
        };
        FeatureVersion.prototype.toObject = function () {
            return {
                id: this.id,
                name: this.name,
                version: this.version,
                value: this.value,
                strategies: this.strategies.toObject(),
                extendField: this.extendField.toObject(),
                isAbtFirst: this.isAbtFirst,
                stgRelation: this.stgRelation,
                timeLimitType: this.timeLimitType,
                timeLimits: this.timeLimits.toObject(),
                refreshType: this.refreshType,
                prerequisites: this.prerequisites.toObject(),
                strategyGroups: this.strategyGroups.toObject(),
                defaultDatasetEx: this.defaultDatasetEx.toObject(),
                offDatasetEx: this.offDatasetEx.toObject(),
                flowStatus: this.flowStatus.toObject()
            };
        };
        FeatureVersion.prototype.readFromObject = function (json) {
            _hasOwnProperty.call(json, "id") && (this.id = json.id);
            _hasOwnProperty.call(json, "name") && (this.name = json.name);
            _hasOwnProperty.call(json, "version") && (this.version = json.version);
            _hasOwnProperty.call(json, "value") && (this.value = json.value);
            _hasOwnProperty.call(json, "strategies") && (this.strategies.readFromObject(json.strategies));
            _hasOwnProperty.call(json, "extendField") && (this.extendField.readFromObject(json.extendField));
            _hasOwnProperty.call(json, "isAbtFirst") && (this.isAbtFirst = json.isAbtFirst);
            _hasOwnProperty.call(json, "stgRelation") && (this.stgRelation = json.stgRelation);
            _hasOwnProperty.call(json, "timeLimitType") && (this.timeLimitType = json.timeLimitType);
            _hasOwnProperty.call(json, "timeLimits") && (this.timeLimits.readFromObject(json.timeLimits));
            _hasOwnProperty.call(json, "refreshType") && (this.refreshType = json.refreshType);
            _hasOwnProperty.call(json, "prerequisites") && (this.prerequisites.readFromObject(json.prerequisites));
            _hasOwnProperty.call(json, "strategyGroups") && (this.strategyGroups.readFromObject(json.strategyGroups));
            _hasOwnProperty.call(json, "defaultDatasetEx") && (this.defaultDatasetEx.readFromObject(json.defaultDatasetEx));
            _hasOwnProperty.call(json, "offDatasetEx") && (this.offDatasetEx.readFromObject(json.offDatasetEx));
            _hasOwnProperty.call(json, "flowStatus") && (this.flowStatus.readFromObject(json.flowStatus));
            return this;
        };
        FeatureVersion.prototype.toBinBuffer = function () {
            var os = new TafStream.JceOutputStream();
            this._writeTo(os);
            return os.getBinBuffer();
        };
        FeatureVersion.new = function () {
            return new FeatureVersion();
        };
        FeatureVersion.create = function (is) {
            return dto.FeatureVersion._readFrom(is);
        };
        FeatureVersion._classname = "dto.FeatureVersion";
        return FeatureVersion;
    }());
    dto.FeatureVersion = FeatureVersion;
    var QueryFeatureWithStrategyResp = /** @class */ (function () {
        function QueryFeatureWithStrategyResp() {
            this.code = 0;
            this.msg = "";
            this.nextTime = 0;
            this.setName = "";
            this.data = new TafStream.List(dto.FeatureVersion);
            this.timestamp = 0;
            this.deletedFeatures = new TafStream.List(TafStream.Int32);
            this.increment = true;
            this.serverTime = 0;
            this.prerequisites = new TafStream.List(dto.FeatureVersion);
            this._proto_struct_name_ = "";
            this._classname = "dto.QueryFeatureWithStrategyResp";
        }
        QueryFeatureWithStrategyResp._write = function (os, tag, val) { os.writeStruct(tag, val); };
        QueryFeatureWithStrategyResp._read = function (is, tag, def) { return is.readStruct(tag, true, def); };
        QueryFeatureWithStrategyResp._readFrom = function (is) {
            var tmp = new QueryFeatureWithStrategyResp;
            tmp.code = is.readInt32(0, true, 0);
            tmp.msg = is.readString(1, true, "");
            tmp.nextTime = is.readInt32(2, false, 0);
            tmp.setName = is.readString(3, false, "");
            tmp.data = is.readList(4, false, TafStream.List(dto.FeatureVersion));
            tmp.timestamp = is.readInt64(5, false, 0);
            tmp.deletedFeatures = is.readList(6, false, TafStream.List(TafStream.Int32));
            tmp.increment = is.readBoolean(7, false, true);
            tmp.serverTime = is.readInt64(8, false, 0);
            tmp.prerequisites = is.readList(9, false, TafStream.List(dto.FeatureVersion));
            return tmp;
        };
        QueryFeatureWithStrategyResp.prototype._writeTo = function (os) {
            os.writeInt32(0, this.code);
            os.writeString(1, this.msg);
            os.writeInt32(2, this.nextTime);
            os.writeString(3, this.setName);
            os.writeList(4, this.data);
            os.writeInt64(5, this.timestamp);
            os.writeList(6, this.deletedFeatures);
            os.writeBoolean(7, this.increment);
            os.writeInt64(8, this.serverTime);
            os.writeList(9, this.prerequisites);
        };
        QueryFeatureWithStrategyResp.prototype._equal = function () {
            assert.fail("this structure not define key operation");
        };
        QueryFeatureWithStrategyResp.prototype._genKey = function () {
            if (!this._proto_struct_name_) {
                this._proto_struct_name_ = "STRUCT" + Math.random();
            }
            return this._proto_struct_name_;
        };
        QueryFeatureWithStrategyResp.prototype.toObject = function () {
            return {
                code: this.code,
                msg: this.msg,
                nextTime: this.nextTime,
                setName: this.setName,
                data: this.data.toObject(),
                timestamp: this.timestamp,
                deletedFeatures: this.deletedFeatures.toObject(),
                increment: this.increment,
                serverTime: this.serverTime,
                prerequisites: this.prerequisites.toObject()
            };
        };
        QueryFeatureWithStrategyResp.prototype.readFromObject = function (json) {
            _hasOwnProperty.call(json, "code") && (this.code = json.code);
            _hasOwnProperty.call(json, "msg") && (this.msg = json.msg);
            _hasOwnProperty.call(json, "nextTime") && (this.nextTime = json.nextTime);
            _hasOwnProperty.call(json, "setName") && (this.setName = json.setName);
            _hasOwnProperty.call(json, "data") && (this.data.readFromObject(json.data));
            _hasOwnProperty.call(json, "timestamp") && (this.timestamp = json.timestamp);
            _hasOwnProperty.call(json, "deletedFeatures") && (this.deletedFeatures.readFromObject(json.deletedFeatures));
            _hasOwnProperty.call(json, "increment") && (this.increment = json.increment);
            _hasOwnProperty.call(json, "serverTime") && (this.serverTime = json.serverTime);
            _hasOwnProperty.call(json, "prerequisites") && (this.prerequisites.readFromObject(json.prerequisites));
            return this;
        };
        QueryFeatureWithStrategyResp.prototype.toBinBuffer = function () {
            var os = new TafStream.JceOutputStream();
            this._writeTo(os);
            return os.getBinBuffer();
        };
        QueryFeatureWithStrategyResp.new = function () {
            return new QueryFeatureWithStrategyResp();
        };
        QueryFeatureWithStrategyResp.create = function (is) {
            return dto.QueryFeatureWithStrategyResp._readFrom(is);
        };
        QueryFeatureWithStrategyResp._classname = "dto.QueryFeatureWithStrategyResp";
        return QueryFeatureWithStrategyResp;
    }());
    dto.QueryFeatureWithStrategyResp = QueryFeatureWithStrategyResp;
    var QueryProductSetReq = /** @class */ (function () {
        function QueryProductSetReq() {
            this.productId = "";
            this._proto_struct_name_ = "";
            this._classname = "dto.QueryProductSetReq";
        }
        QueryProductSetReq._write = function (os, tag, val) { os.writeStruct(tag, val); };
        QueryProductSetReq._read = function (is, tag, def) { return is.readStruct(tag, true, def); };
        QueryProductSetReq._readFrom = function (is) {
            var tmp = new QueryProductSetReq;
            tmp.productId = is.readString(0, true, "");
            return tmp;
        };
        QueryProductSetReq.prototype._writeTo = function (os) {
            os.writeString(0, this.productId);
        };
        QueryProductSetReq.prototype._equal = function () {
            assert.fail("this structure not define key operation");
        };
        QueryProductSetReq.prototype._genKey = function () {
            if (!this._proto_struct_name_) {
                this._proto_struct_name_ = "STRUCT" + Math.random();
            }
            return this._proto_struct_name_;
        };
        QueryProductSetReq.prototype.toObject = function () {
            return {
                productId: this.productId
            };
        };
        QueryProductSetReq.prototype.readFromObject = function (json) {
            _hasOwnProperty.call(json, "productId") && (this.productId = json.productId);
            return this;
        };
        QueryProductSetReq.prototype.toBinBuffer = function () {
            var os = new TafStream.JceOutputStream();
            this._writeTo(os);
            return os.getBinBuffer();
        };
        QueryProductSetReq.new = function () {
            return new QueryProductSetReq();
        };
        QueryProductSetReq.create = function (is) {
            return dto.QueryProductSetReq._readFrom(is);
        };
        QueryProductSetReq._classname = "dto.QueryProductSetReq";
        return QueryProductSetReq;
    }());
    dto.QueryProductSetReq = QueryProductSetReq;
    var AccessSet = /** @class */ (function () {
        function AccessSet() {
            this.setName = "";
            this.l5ModId = "";
            this.l5CmdId = "";
            this._proto_struct_name_ = "";
            this._classname = "dto.AccessSet";
        }
        AccessSet._write = function (os, tag, val) { os.writeStruct(tag, val); };
        AccessSet._read = function (is, tag, def) { return is.readStruct(tag, true, def); };
        AccessSet._readFrom = function (is) {
            var tmp = new AccessSet;
            tmp.setName = is.readString(0, false, "");
            tmp.l5ModId = is.readString(1, false, "");
            tmp.l5CmdId = is.readString(2, false, "");
            return tmp;
        };
        AccessSet.prototype._writeTo = function (os) {
            os.writeString(0, this.setName);
            os.writeString(1, this.l5ModId);
            os.writeString(2, this.l5CmdId);
        };
        AccessSet.prototype._equal = function () {
            assert.fail("this structure not define key operation");
        };
        AccessSet.prototype._genKey = function () {
            if (!this._proto_struct_name_) {
                this._proto_struct_name_ = "STRUCT" + Math.random();
            }
            return this._proto_struct_name_;
        };
        AccessSet.prototype.toObject = function () {
            return {
                setName: this.setName,
                l5ModId: this.l5ModId,
                l5CmdId: this.l5CmdId
            };
        };
        AccessSet.prototype.readFromObject = function (json) {
            _hasOwnProperty.call(json, "setName") && (this.setName = json.setName);
            _hasOwnProperty.call(json, "l5ModId") && (this.l5ModId = json.l5ModId);
            _hasOwnProperty.call(json, "l5CmdId") && (this.l5CmdId = json.l5CmdId);
            return this;
        };
        AccessSet.prototype.toBinBuffer = function () {
            var os = new TafStream.JceOutputStream();
            this._writeTo(os);
            return os.getBinBuffer();
        };
        AccessSet.new = function () {
            return new AccessSet();
        };
        AccessSet.create = function (is) {
            return dto.AccessSet._readFrom(is);
        };
        AccessSet._classname = "dto.AccessSet";
        return AccessSet;
    }());
    dto.AccessSet = AccessSet;
    var QueryProductSetResp = /** @class */ (function () {
        function QueryProductSetResp() {
            this.code = 0;
            this.msg = "";
            this.data = new dto.AccessSet;
            this._proto_struct_name_ = "";
            this._classname = "dto.QueryProductSetResp";
        }
        QueryProductSetResp._write = function (os, tag, val) { os.writeStruct(tag, val); };
        QueryProductSetResp._read = function (is, tag, def) { return is.readStruct(tag, true, def); };
        QueryProductSetResp._readFrom = function (is) {
            var tmp = new QueryProductSetResp;
            tmp.code = is.readInt32(0, true, 0);
            tmp.msg = is.readString(1, true, "");
            tmp.data = is.readStruct(2, false, dto.AccessSet);
            return tmp;
        };
        QueryProductSetResp.prototype._writeTo = function (os) {
            os.writeInt32(0, this.code);
            os.writeString(1, this.msg);
            os.writeStruct(2, this.data);
        };
        QueryProductSetResp.prototype._equal = function () {
            assert.fail("this structure not define key operation");
        };
        QueryProductSetResp.prototype._genKey = function () {
            if (!this._proto_struct_name_) {
                this._proto_struct_name_ = "STRUCT" + Math.random();
            }
            return this._proto_struct_name_;
        };
        QueryProductSetResp.prototype.toObject = function () {
            return {
                code: this.code,
                msg: this.msg,
                data: this.data.toObject()
            };
        };
        QueryProductSetResp.prototype.readFromObject = function (json) {
            _hasOwnProperty.call(json, "code") && (this.code = json.code);
            _hasOwnProperty.call(json, "msg") && (this.msg = json.msg);
            _hasOwnProperty.call(json, "data") && (this.data.readFromObject(json.data));
            return this;
        };
        QueryProductSetResp.prototype.toBinBuffer = function () {
            var os = new TafStream.JceOutputStream();
            this._writeTo(os);
            return os.getBinBuffer();
        };
        QueryProductSetResp.new = function () {
            return new QueryProductSetResp();
        };
        QueryProductSetResp.create = function (is) {
            return dto.QueryProductSetResp._readFrom(is);
        };
        QueryProductSetResp._classname = "dto.QueryProductSetResp";
        return QueryProductSetResp;
    }());
    dto.QueryProductSetResp = QueryProductSetResp;
    var FeatureTriggerEvent = /** @class */ (function () {
        function FeatureTriggerEvent() {
            this.sn = "";
            this.featureVersionId = "";
            this.featureName = "";
            this.featureValue = "";
            this.abTestId = "-1";
            this.userId = "";
            this.eventCount = 0;
            this.qimei = "";
            this.datasetID = "";
            this.versionID = "";
            this._proto_struct_name_ = "";
            this._classname = "dto.FeatureTriggerEvent";
        }
        FeatureTriggerEvent._write = function (os, tag, val) { os.writeStruct(tag, val); };
        FeatureTriggerEvent._read = function (is, tag, def) { return is.readStruct(tag, true, def); };
        FeatureTriggerEvent._readFrom = function (is) {
            var tmp = new FeatureTriggerEvent;
            tmp.sn = is.readString(0, true, "");
            tmp.featureVersionId = is.readString(1, true, "");
            tmp.featureName = is.readString(2, true, "");
            tmp.featureValue = is.readString(3, true, "");
            tmp.abTestId = is.readString(4, true, "-1");
            tmp.userId = is.readString(5, true, "");
            tmp.eventCount = is.readInt64(6, true, 0);
            tmp.qimei = is.readString(7, true, "");
            tmp.datasetID = is.readString(8, true, "");
            tmp.versionID = is.readString(9, true, "");
            return tmp;
        };
        FeatureTriggerEvent.prototype._writeTo = function (os) {
            os.writeString(0, this.sn);
            os.writeString(1, this.featureVersionId);
            os.writeString(2, this.featureName);
            os.writeString(3, this.featureValue);
            os.writeString(4, this.abTestId);
            os.writeString(5, this.userId);
            os.writeInt64(6, this.eventCount);
            os.writeString(7, this.qimei);
            os.writeString(8, this.datasetID);
            os.writeString(9, this.versionID);
        };
        FeatureTriggerEvent.prototype._equal = function () {
            assert.fail("this structure not define key operation");
        };
        FeatureTriggerEvent.prototype._genKey = function () {
            if (!this._proto_struct_name_) {
                this._proto_struct_name_ = "STRUCT" + Math.random();
            }
            return this._proto_struct_name_;
        };
        FeatureTriggerEvent.prototype.toObject = function () {
            return {
                sn: this.sn,
                featureVersionId: this.featureVersionId,
                featureName: this.featureName,
                featureValue: this.featureValue,
                abTestId: this.abTestId,
                userId: this.userId,
                eventCount: this.eventCount,
                qimei: this.qimei,
                datasetID: this.datasetID,
                versionID: this.versionID
            };
        };
        FeatureTriggerEvent.prototype.readFromObject = function (json) {
            _hasOwnProperty.call(json, "sn") && (this.sn = json.sn);
            _hasOwnProperty.call(json, "featureVersionId") && (this.featureVersionId = json.featureVersionId);
            _hasOwnProperty.call(json, "featureName") && (this.featureName = json.featureName);
            _hasOwnProperty.call(json, "featureValue") && (this.featureValue = json.featureValue);
            _hasOwnProperty.call(json, "abTestId") && (this.abTestId = json.abTestId);
            _hasOwnProperty.call(json, "userId") && (this.userId = json.userId);
            _hasOwnProperty.call(json, "eventCount") && (this.eventCount = json.eventCount);
            _hasOwnProperty.call(json, "qimei") && (this.qimei = json.qimei);
            _hasOwnProperty.call(json, "datasetID") && (this.datasetID = json.datasetID);
            _hasOwnProperty.call(json, "versionID") && (this.versionID = json.versionID);
            return this;
        };
        FeatureTriggerEvent.prototype.toBinBuffer = function () {
            var os = new TafStream.JceOutputStream();
            this._writeTo(os);
            return os.getBinBuffer();
        };
        FeatureTriggerEvent.new = function () {
            return new FeatureTriggerEvent();
        };
        FeatureTriggerEvent.create = function (is) {
            return dto.FeatureTriggerEvent._readFrom(is);
        };
        FeatureTriggerEvent._classname = "dto.FeatureTriggerEvent";
        return FeatureTriggerEvent;
    }());
    dto.FeatureTriggerEvent = FeatureTriggerEvent;
    var ReportFeatureTriggerEventReq = /** @class */ (function () {
        function ReportFeatureTriggerEventReq() {
            this.sdkType = "";
            this.sdkVersion = "";
            this.startTime = "";
            this.stopTime = "";
            this.productId = "";
            this.moduleId = "";
            this.moduleVersion = "";
            this.events = new TafStream.List(dto.FeatureTriggerEvent);
            this.bundleId = "";
            this.qua = "";
            this.channel = "";
            this.os = "";
            this.osVersion = "";
            this.deviceId = "";
            this.uploadIp = "";
            this._proto_struct_name_ = "";
            this._classname = "dto.ReportFeatureTriggerEventReq";
        }
        ReportFeatureTriggerEventReq._write = function (os, tag, val) { os.writeStruct(tag, val); };
        ReportFeatureTriggerEventReq._read = function (is, tag, def) { return is.readStruct(tag, true, def); };
        ReportFeatureTriggerEventReq._readFrom = function (is) {
            var tmp = new ReportFeatureTriggerEventReq;
            tmp.sdkType = is.readString(0, true, "");
            tmp.sdkVersion = is.readString(1, true, "");
            tmp.startTime = is.readString(2, true, "");
            tmp.stopTime = is.readString(3, true, "");
            tmp.productId = is.readString(4, true, "");
            tmp.moduleId = is.readString(5, true, "");
            tmp.moduleVersion = is.readString(6, true, "");
            tmp.events = is.readList(7, true, TafStream.List(dto.FeatureTriggerEvent));
            tmp.bundleId = is.readString(8, false, "");
            tmp.qua = is.readString(9, false, "");
            tmp.channel = is.readString(10, false, "");
            tmp.os = is.readString(11, false, "");
            tmp.osVersion = is.readString(12, false, "");
            tmp.deviceId = is.readString(13, false, "");
            tmp.uploadIp = is.readString(14, false, "");
            return tmp;
        };
        ReportFeatureTriggerEventReq.prototype._writeTo = function (os) {
            os.writeString(0, this.sdkType);
            os.writeString(1, this.sdkVersion);
            os.writeString(2, this.startTime);
            os.writeString(3, this.stopTime);
            os.writeString(4, this.productId);
            os.writeString(5, this.moduleId);
            os.writeString(6, this.moduleVersion);
            os.writeList(7, this.events);
            os.writeString(8, this.bundleId);
            os.writeString(9, this.qua);
            os.writeString(10, this.channel);
            os.writeString(11, this.os);
            os.writeString(12, this.osVersion);
            os.writeString(13, this.deviceId);
            os.writeString(14, this.uploadIp);
        };
        ReportFeatureTriggerEventReq.prototype._equal = function () {
            assert.fail("this structure not define key operation");
        };
        ReportFeatureTriggerEventReq.prototype._genKey = function () {
            if (!this._proto_struct_name_) {
                this._proto_struct_name_ = "STRUCT" + Math.random();
            }
            return this._proto_struct_name_;
        };
        ReportFeatureTriggerEventReq.prototype.toObject = function () {
            return {
                sdkType: this.sdkType,
                sdkVersion: this.sdkVersion,
                startTime: this.startTime,
                stopTime: this.stopTime,
                productId: this.productId,
                moduleId: this.moduleId,
                moduleVersion: this.moduleVersion,
                events: this.events.toObject(),
                bundleId: this.bundleId,
                qua: this.qua,
                channel: this.channel,
                os: this.os,
                osVersion: this.osVersion,
                deviceId: this.deviceId,
                uploadIp: this.uploadIp
            };
        };
        ReportFeatureTriggerEventReq.prototype.readFromObject = function (json) {
            _hasOwnProperty.call(json, "sdkType") && (this.sdkType = json.sdkType);
            _hasOwnProperty.call(json, "sdkVersion") && (this.sdkVersion = json.sdkVersion);
            _hasOwnProperty.call(json, "startTime") && (this.startTime = json.startTime);
            _hasOwnProperty.call(json, "stopTime") && (this.stopTime = json.stopTime);
            _hasOwnProperty.call(json, "productId") && (this.productId = json.productId);
            _hasOwnProperty.call(json, "moduleId") && (this.moduleId = json.moduleId);
            _hasOwnProperty.call(json, "moduleVersion") && (this.moduleVersion = json.moduleVersion);
            _hasOwnProperty.call(json, "events") && (this.events.readFromObject(json.events));
            _hasOwnProperty.call(json, "bundleId") && (this.bundleId = json.bundleId);
            _hasOwnProperty.call(json, "qua") && (this.qua = json.qua);
            _hasOwnProperty.call(json, "channel") && (this.channel = json.channel);
            _hasOwnProperty.call(json, "os") && (this.os = json.os);
            _hasOwnProperty.call(json, "osVersion") && (this.osVersion = json.osVersion);
            _hasOwnProperty.call(json, "deviceId") && (this.deviceId = json.deviceId);
            _hasOwnProperty.call(json, "uploadIp") && (this.uploadIp = json.uploadIp);
            return this;
        };
        ReportFeatureTriggerEventReq.prototype.toBinBuffer = function () {
            var os = new TafStream.JceOutputStream();
            this._writeTo(os);
            return os.getBinBuffer();
        };
        ReportFeatureTriggerEventReq.new = function () {
            return new ReportFeatureTriggerEventReq();
        };
        ReportFeatureTriggerEventReq.create = function (is) {
            return dto.ReportFeatureTriggerEventReq._readFrom(is);
        };
        ReportFeatureTriggerEventReq._classname = "dto.ReportFeatureTriggerEventReq";
        return ReportFeatureTriggerEventReq;
    }());
    dto.ReportFeatureTriggerEventReq = ReportFeatureTriggerEventReq;
})(dto = exports.dto || (exports.dto = {}));
