// @tencent/toggle-server-node/lib/storage.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.Storage = void 0;
var events_1 = require("events");
var path_1 = require("path");
var fs_1 = require("fs");
var Storage = /** @class */ (function (_super) {
    __extends(Storage, _super);
    function Storage(_a) {
        var backupPath = _a.backupPath, productEn = _a.productEn, moduleId = _a.moduleId;
        var _this = _super.call(this) || this;
        // ready is a "ready"-flag to signal that storage is ready with data,
        // and to signal to backup not to store fetched backup
        _this.ready = false;
        _this.data = {
            featureData: {},
            prerequisite: {},
        };
        _this.path = path_1.join(backupPath, "/toggle-repo-schema-v1-" + productEn + "-" + moduleId + ".json");
        _this.backupDir = path_1.dirname(_this.path);
        _this.load();
        return _this;
    }
    Storage.prototype.reset = function (data, doPersist) {
        var _this = this;
        if (doPersist === void 0) { doPersist = true; }
        var doEmitReady = this.ready === false;
        this.ready = true;
        this.data = data;
        process.nextTick(function () {
            if (doEmitReady) {
                _this.emit('ready');
            }
            if (doPersist) {
                _this.persist();
            }
        });
    };
    Storage.prototype.get = function (key) {
        return this.data.featureData[key];
    };
    Storage.prototype.getPrerequisite = function (key) {
        return this.data.prerequisite[key];
    };
    Storage.prototype.getAll = function () {
        return this.data.featureData;
    };
    Storage.prototype.persist = function () {
        var _this = this;
        fs_1.mkdir(this.backupDir, { recursive: true }, function (err) {
            if (err) {
                return _this.emit('warn', "create dir " + _this.backupDir + " failed. err is " + err);
            }
            fs_1.writeFile(_this.path, JSON.stringify(_this.data), function (err) {
                if (err) {
                    return _this.emit('warn', err);
                }
                _this.emit('persisted', true);
            });
        });
    };
    Storage.prototype.load = function () {
        var _this = this;
        fs_1.readFile(this.path, 'utf8', function (err, data) {
            if (_this.ready) {
                return;
            }
            if (err) {
                if (err.code !== 'ENOENT') {
                    _this.emit('warn', err);
                }
                return;
            }
            try {
                _this.reset(JSON.parse(data), false);
            }
            catch (err) {
                err.message = "Toggle storage failed parsing file " + _this.path + ": " + err.message;
                _this.emit('warn', err);
            }
        });
    };
    Storage.prototype.incUpdate = function (featureToUpdate, newPrerequisite, featureSimpleMap, featureToDelete, doPersist) {
        if (doPersist === void 0) { doPersist = true; }
        if (!this.data || Object.keys(this.data.featureData).length === 0) {
            this.data.featureData = featureToUpdate;
            this.data.prerequisite = newPrerequisite;
        }
        else {
            // add or update
            for (var key in featureToUpdate) {
                this.data.featureData[key] = featureToUpdate[key];
            }
            // delete
            for (var _i = 0, featureToDelete_1 = featureToDelete; _i < featureToDelete_1.length; _i++) {
                var id = featureToDelete_1[_i];
                delete this.data.featureData[featureSimpleMap[id] || ''];
            }
        }
        this.data.prerequisite = newPrerequisite;
        if (doPersist) {
            this.persist();
        }
    };
    return Storage;
}(events_1.EventEmitter));
exports.Storage = Storage;
